File: lib/features/payments/domain/payment_status_machine.dart
import '../../shared/domain/enums.dart';

abstract final class PaymentStatusMachine {
  static const Map<PaymentStatus, Set<PaymentStatus>> _transitions = {
    PaymentStatus.unpaid: {
      PaymentStatus.paymentPending,
      PaymentStatus.paid,
    },
    PaymentStatus.paymentPending: {PaymentStatus.paid},
    PaymentStatus.paid: {},
  };

  static bool canTransition(PaymentStatus from, PaymentStatus to) {
    return _transitions[from]?.contains(to) ?? false;
  }

  static bool canMarkPaid(PaymentStatus current) {
    return canTransition(current, PaymentStatus.paid);
  }

  static void assertTransition(PaymentStatus from, PaymentStatus to) {
    if (!canTransition(from, to)) {
      throw StateError('Invalid payment status transition: $from → $to');
    }
  }
}

File: backend/app/Services/PaymentStatusMachine.php
<?php

namespace App\Services;

use App\Enums\PaymentStatus;
use InvalidArgumentException;

class PaymentStatusMachine
{
    /**
     * @var array<string, list<string>>
     */
    private const TRANSITIONS = [
        PaymentStatus::Unpaid->value => [
            PaymentStatus::PaymentPending->value,
            PaymentStatus::Paid->value,
        ],
        PaymentStatus::PaymentPending->value => [
            PaymentStatus::Paid->value,
            PaymentStatus::Unpaid->value,
        ],
        PaymentStatus::Paid->value => [],
    ];

    public function canTransition(PaymentStatus $from, PaymentStatus $to): bool
    {
        $allowed = self::TRANSITIONS[$from->value] ?? [];

        return in_array($to->value, $allowed, true);
    }

    public function assertCanTransition(PaymentStatus $from, PaymentStatus $to): void
    {
        if (! $this->canTransition($from, $to)) {
            throw new InvalidArgumentException(
                "Cannot transition payment from {$from->value} to {$to->value}."
            );
        }
    }
}
